### DISCUSSION TITLE
Can't authenticate Data Importer with Firefly III

### DISCUSSION THREAD

**Discussion by 2ndwonder** [firefly-iii/firefly-iii#8654] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

Can't authenticate Data Importer with Firefly III

### Debug information

Debug information generated at 2024-03-10 16:00:24 Europe/Amsterdam for Firefly III Data Importer version **1.4.4**.

<table><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr><tr><td>Version</td><td>1.4.4</td></tr><tr><td>Build</td><td><span>#</span>, base <span>#</span>80</td></tr><tr><td>System</td><td>PHP 8.3.3, Linux, apache2handler</td></tr></table><table><tr><th colspan="2">App information</th></tr><tr><th>Item</th><th>Value</th></tr><tr><td>Timezone</td><td>Europe/Amsterdam, Europe/Amsterdam</td></tr><tr><td>Environment</td><td>local</td></tr><tr><td>Debug mode</td><td>false, cache 'file'</td></tr><tr><td>Log level</td><td>debug, stack</td></tr><tr><td>Display errors</td><td>Off, ALL errors</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Trusted proxies</td><td></td></tr></table><table><tr><th colspan="2">User information</th></tr><tr><th>Item</th><th>Value</th></tr><tr><td>User agent</td><td>Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.3.1 Safari/605.1.15</td></tr></table>

### Expected behaviour

When I try to authenticate Data Importer with Firefly III I get a 500 error.

The error occurred in /var/www/html/app/Http/Controllers/TokenController.php:71.

I'm on a Synology NAS and using Docker-compose.

### Steps to reproduce

I configured everything in my .env files and Firefly III works without any issues. Only problem is Data Importer which cannot authenticate with Firefly III 

### Additional info

None

**Comment by 2ndwonder** [comment]
Can anyone please assist?

**Comment by InSelfControll** [comment]
Hey I got the same issue can't authenticate data importer..
Debug information generated at 2024-03-12 12:03:15 Asia/Jerusalem for Firefly III Data Importer version **1.4.4**.

<table><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr><tr><td>Version</td><td>1.4.4</td></tr><tr><td>Build</td><td><span>#</span>, base <span>#</span>80</td></tr><tr><td>System</td><td>PHP 8.3.3, Linux, apache2handler</td></tr></table><table><tr><th colspan="2">App information</th></tr><tr><th>Item</th><th>Value</th></tr><tr><td>Timezone</td><td>Europe/Amsterdam, Asia/Jerusalem</td></tr><tr><td>Environment</td><td>production</td></tr><tr><td>Debug mode</td><td>false, cache 'file'</td></tr><tr><td>Log level</td><td>debug, stack</td></tr><tr><td>Display errors</td><td>Off, ALL errors</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Trusted proxies</td><td></td></tr></table><table><tr><th colspan="2">User information</th></tr><tr><th>Item</th><th>Value</th></tr><tr><td>User agent</td><td>Mozilla/5.0 (X11; Linux x86_64; rv:120.0) Gecko/20100101 Firefox/120.0</td></tr></table>

Expected behaviour

When I try to authenticate Data Importer with Firefly III I get a 500 error.

The error occurred in /var/www/html/app/Http/Controllers/TokenController.php:71.

I'm on a Podman VPS with cloudflared tunnel and using Docker-compose.
Steps to reproduce

I configured everything in my environment vars in portainer Firefly III works without any issues. Only problem is Data Importer which cannot authenticate with Firefly III

**Comment by JC5** [comment]
Could you tell me what the error is?

**Reply by InSelfControll** [reply]
![image](https://github.com/firefly-iii/firefly-iii/assets/20114252/15e8ba17-be77-42e3-8f57-55ba1cf92cbe)

**Reply by JC5** [reply]
Are you approaching either the data importer or Firefly III by IP?

**Reply by InSelfControll** [reply]
I tried both with IP and domain but both didn't work.
I get it when trying to authorize the access from the importer to the main system firefly-iii with oauth2.

**Reply by JC5** [reply]
The data importer sets a cookie with the state and sends it to Firefly III. Firefly III should redirect you with the same state. In your logs you can see which state was expected and what was sent back.

**Reply by InSelfControll** [reply]
I'm behind cloudflared so the app can't access it via http only https this is the only error I get. Error 500

**Reply by InSelfControll** [reply]
I think I found the problem.
The importer doesn't get the access token and the client id in the request.
Even when I pass it from the importer interface by adding the url of the app and the client id

**Reply by 2ndwonder** [reply]
Ok.. so how are we going to fix this 😁 I regret buying a Synology NAS now..

**Reply by InSelfControll** [reply]
Please make sure you're using PAT access token for the importer as the first step.
Now I'm facing an issue with cloudflare redirects to https the second domain for importer.
With https it doesn't work at all can't even access the callback and without https it's working but still can't be authorized with firefly iii for some reason.
@JC5

**Reply by InSelfControll** [reply]
@2ndwonder access both of the systems by their IP address directly and not from the domain with SSL certificate only http protocol.
That's the only way I succeed in getting it to work.

**Reply by 2ndwonder** [reply]
Yeah.. that's gonna be a problem. I use tailscale. My NAS is not located in my home. But another friend tried it with tailscale and it works on his side. I think it's a Synology thing

**Reply by JC5** [reply]
All I can add from my end is that I often see that browsers have trouble with cookies when the website is accessed from an IP, and not a (sub)domain name. So if you get redirected back to 10.10.10.whatever, the browser will not use cookies or store cookies. 

That's when it loses the state, and that's the error you're getting.

**Reply by InSelfControll** [reply]
> All I can add from my end is that I often see that browsers have trouble with cookies when the website is accessed from an IP, and not a (sub)domain name. So if you get redirected back to 10.10.10.whatever, the browser will not use cookies or store cookies. 
> 
> That's when it loses the state, and that's the error you're getting.

At the beginning I tried with subdomain but it never worked because the SSL then I tried with IP on both sides and it works but the importer doesn't do what I thought it would do.

Need to find a way to sync my data live from my bank for being able to monitor everything live.

Thanks anyway :)

**Reply by InSelfControll** [reply]
> Yeah.. that's gonna be a problem. I use tailscale. My NAS is not located in my home. But another friend tried it with tailscale and it works on his side. I think it's a Synology thing 

I'm behind a cloudflared tunnel instead of a tailscale and both of my servers are connected to each other via tailscale.

My portainer connected to his agent behind the tailscale.
 What is the error you get?

**Reply by 2ndwonder** [reply]
500 error

I'm running Firefly only through tailscale. 

The error occurred in /var/www/html/app/Http/Controllers/TokenController.php:71.

**Comment by 2ndwonder** [comment]
> > Yeah.. that's gonna be a problem. I use tailscale. My NAS is not located in my home. But another friend tried it with tailscale and it works on his side. I think it's a Synology thing
> 
> I'm behind a cloudflared tunnel instead of a tailscale and both of my servers are connected to each other via tailscale.
> 
> My portainer connected to his agent behind the tailscale. What is the error you get?

Did you figure it out? Importer still doesn't work out for me

**Reply by InSelfControll** [reply]
The importer just doesn't support https connect it via the IP and without https
